File: tests/07_phases_ompss.dg/fortran/success_task_07.f90
! <testinfo>
! test_generator=config/mercurium-ompss
! </testinfo>
PROGRAM MAIN
  IMPLICIT NONE

  INTEGER :: A, B
  INTEGER, DIMENSION(:), ALLOCATABLE :: X, Y

  ALLOCATE(X(2:100), Y(2:100))

  A = 5
  B = 10

  Y = 42

  !$OMP TASK OUT(X(A:B))
  X(5:10) = 1
  !$OMP END TASK

  A = 5+1
  B = 10+1

  !$OMP TASK IN(X(A-1:B-1)) OUT(Y(A-1:B-1))
  Y(5:10) = X(5:10) + 3
  !$OMP END TASK

  !$OMP TASK IN(Y(A-1:B-1))
  IF (ANY(Y(5:10) /= 4)) STOP 1
  !$OMP END TASK

  !$OMP TASKWAIT

  IF (ANY(Y(:4) /= 42)) STOP 2
  IF (ANY(Y(11:) /= 42)) STOP 3
END PROGRAM MAIN

File: tests/07_phases_ompss.dg/cxx/success_task_07.cpp
/*--------------------------------------------------------------------
  (C) Copyright 2006-2012 Barcelona Supercomputing Center
                          Centro Nacional de Supercomputacion
  
  This file is part of Mercurium C/C++ source-to-source compiler.
  
  See AUTHORS file in the top level directory for information
  regarding developers and contributors.
  
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.
  
  Mercurium C/C++ source-to-source compiler is distributed in the hope
  that it will be useful, but WITHOUT ANY WARRANTY; without even the
  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
  PURPOSE.  See the GNU Lesser General Public License for more
  details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with Mercurium C/C++ source-to-source compiler; if
  not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.
--------------------------------------------------------------------*/



/*
<testinfo>
test_generator=config/mercurium-ompss
</testinfo>
*/
#include <stdlib.h>

/* test.cpp */
void foo(int* a, int n)
{
#pragma omp task inout(a[0;n])
    {
        int i;
        for (i = 0; i < n; i++)
        {
            a[i]++;
        }
    }
#pragma omp taskwait
}

int main(int argc, char *argv[])
{
    const int t[10] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    int x[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    int *ptr = x;

    foo(ptr, 10);

    int i;
    for (i = 0; i < 10; i++)
    {
        if (x[i] != t[i])
            abort();
    }
}
